### DISCUSSION TITLE
Servlet 请求的回放过程

### DISCUSSION THREAD

**Discussion by buzz321** [arextest/arex-agent-java#500] [Q&A] [answered]
1、通过 arex-web 进行 Servlet 请求的回放时，目标机器会生成 4 次arex-replay-id，
![image](https://github.com/arextest/arex-agent-java/assets/77421340/da344290-bb0e-48d9-aa92-a389ce6ecd65)

2、同时调试发现，第 1 次回放开始，执行到 RequestHandlerManager#preHandle 时，REQUEST_HANDLER_CACHE 缓存总为空，因为REQUEST_HANDLER_CACHE 的初始化处理在 RequestHandlerManager#preHandle 后
![581717661284_ pic_hd](https://github.com/arextest/arex-agent-java/assets/77421340/6b42e076-d06b-4533-9da5-901a51f6cc2a)

请问这两个现象是合理的么？

**Comment by mr3** [comment, marked as answer]
@buzz321 
1. 回放流程请求 1次预热请求/2次降噪请求/1...N回放请求
2. 可能与你理解有一些偏差，Request handler 预热流程如下
```java
CaseEventDispatcher.onEvent(CaseEvent.ofEnterEvent());   
    io.arex.inst.runtime.listener.EventProcessor#onRequest
        io.arex.inst.runtime.request.RequestHandlerManager#init
```
详细代码见：https://github.com/arextest/arex-agent-java
